### PR TITLE
Allow non-replicated actors to become replicated ones

### PR DESCRIPTION
**Pull request by jypma**
See individual commit messages.

### CODE REVIEW

File: ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java [hunk 1]
```diff
@@ -20,21 +27,19 @@
  * 
  * The first event emitted by the first command sent to this actor MUST store the local datacenter name, and
  * EventClassifier must return that datacenter name (as first element, if multiple) when given the event.
+ * 
+ * The persistenceId for a replicated actor _must not_ be allowed to be client-picked, since the system must guarantee that
```
Comment (ayurchuk) [inline ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java]: Looks like javadoc uses `<i>` when `*.md` - `_`

**Comment by ayurchuk** [comment]
:star:

File: ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java [hunk 2]
```diff
@@ -46,57 +51,118 @@ public ReplicatedActor(Class<C> commandType, Class<E> eventType) {
...
+    protected Receive migrateNonReplicatedActor() {
+    	self().tell(Migrate.INSTANCE, self()); // we can't use persist() inside receiveRecover(), at least not with the cassandra plugin.
     	
-        return ReceiveBuilder.create()
-            .match(commandType, c -> slave != null && slave && !isReadOnly(c), c ->
-                sender().tell(new Failure(new IllegalStateException("Actor is in slave mode and does not accept " + c)), self())
-            )
-            .match(commandType, c -> slave == null && isReadOnly(c), c -> {
+    	return ReceiveBuilder.create()
+			.match(Migrate.class, msg -> {
```
Comment (alar17) [inline ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java]: I can see tabs here. Could you double check the format?

File: ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/NonReplicatedTestActor.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        return TestActorState.EMPTY;
+    }
+
+    @Override
+    protected PartialFunction<TestCommand, ? extends AbstractCommandHandler<TestCommand, TestEvent, TestActorState>> applyCommand() {
+        return new PFBuilder<TestCommand,Handler>()
+            .match(TestCommand.class, c -> c.hasRead(), c -> new Handler(c) {
+                @Override
+                public Option<Object> getValidationError(long lastSequenceNr) {
+                    if (state.getMsg() == null) {
```
Comment (alar17) [inline ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/NonReplicatedTestActor.java]: Consider changing the msg of TestActorState to Option and avoid checking null

**Comment by alar17** [comment]
@jypma Do you intent to keep all four commits?

**Comment by jypma** [comment]
Yes.

File: ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/NonReplicatedTestActor.java (hunk 3 above)
Comment (jypma) [reply ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/NonReplicatedTestActor.java]: meh. I can live with null in the test.

File: ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java (hunk 2 above)
Comment (jypma) [reply ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java]: thanks! updated.

File: ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/ReplicatedActorSpec.java [hunk 4]
```diff
@@ -170,14 +172,49 @@
...
+    	
+    	it("should emit the extra datacenter event pinning it to the local datacenter on recovery, but only once", () -> {
+            actor.tell(cmd().setRead(TestCommand.Read.newBuilder()).build(), sender.getRef());
+            sender.expectMsg("dc:local"); // because createMigrationEvent() has emitted this.
+            
+            Seq<EventEnvelope> afterFirstRestart = journalEventsFor(actor.path().name());
+            assertThat(TestEvent.class.cast(afterFirstRestart.last().event()).getMsg()).isEqualTo("dc:local");
+
+            // Stop, start, and wait for it to be ready again
+        	sender.watch(actor);
```
Comment (alar17) [inline ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/ReplicatedActorSpec.java]: spacing is not aligned here as well.

Comment (jypma) [reply ts-reaktive-replication/src/test/java/com/tradeshift/reaktive/replication/ReplicatedActorSpec.java]: thanks for spotting. That's my punishment for having under-configured eclipse.

File: ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java [hunk 5]
```diff
@@ -46,57 +51,119 @@ public ReplicatedActor(Class<C> commandType, Class<E> eventType) {
...
+     * The event classifier for E _MUST_ return the local data center as the first datacenter for the returned event.
+     * 
+     * This method will always be called just after recovery has completed, so feel free to use your existing 
+     * state information to create the event, if necessary.
+     * 
+     * By default, this method will throw UnsupportedOperationException.
+     */
+    protected E createMigrationEvent() {
+        throw new UnsupportedOperationException("Trying to migrate an existing non-replicated actor, but createMigrationEvent() wasn't implemented.");
+    }
```
Comment (ayurchuk) [inline ts-reaktive-replication/src/main/java/com/tradeshift/reaktive/replication/actors/ReplicatedActor.java]: Is it true that this method should be implemented by sub classes? Can we then define it as abstract?

**Pull request merged**